### ISSUE TITLE
Request should have Body methods like text(), json(), etc.

### ISSUE THREAD

**Issue by kirill-konshin** [bitinn/node-fetch#36]
`Request` implements `Body`, so it also has the following methods available to it:

- `Body.arrayBuffer()`
- `Body.blob()`
- `Body.formData()`
- `Body.json()`
- `Body.text()`

https://developer.mozilla.org/en-US/docs/Web/API/Request

**Comment by bitinn** [comment]
I am not going to implement this, as I don't see a real need for it. PR is welcomed but I will need to review the use case first.

**Comment by kirill-konshin** [comment]
It's part of the standard... :)

Use case: library gets `Request` object from user, detects that it's content type is `application/json`, get the body, stringifies, returns a new `Request` object with serialized body.

**Comment by bitinn** [comment]
I will quote [my intention](https://github.com/bitinn/node-fetch#features) for this module.

> Make conscious trade-off when following whatwg fetch spec and stream spec implementation details, document known difference.

I don't want to falsely claim `node-fetch` is providing a full API that covers all client-side use case. It would take considerable more code and tests (or dependencies) to avoid messing up, as a lot of client-side API isn't present on node.js

To achieve what you said we at least need to refactor into a `Body` class, and handle things like Blob and formData which doesn't make much sense in a server use-case context.

**Comment by kirill-konshin** [comment]
I'd say, that `Blob` or `FormData` is an obvious overkill. Simple `json()`, `text()` and `buffer()` methods would be more than enough for both `Request` and `Response`.

**Comment by kirill-konshin** [comment]
Especially for given use-case.

**Comment by bitinn** [comment]
Please do show me your code, I feel there are more solutions to your problem.

**Comment by kirill-konshin** [comment]
So far I have a solution -- I simply access `request.body` property. But for truly isomorphic code I'd like to use `text()` and `json()` methods...

Basically, we're talking about sharing part of the code of `Request` for `Response` class as well, as-is, nothing more. Not a big deal, I think.

**Comment by tschaub** [comment]
It looks like `text()` and `json()` were added in af21ae6c1c964cd40e48e974af56ea2e1361304f.

**Comment by bitinn** [comment]
@tschaub nope, he meant Request and Response instances should all inherit Body's methods like `text` and `json`. To me they are an overkill as of now, if you are calling Request ctor directly, you are on the very edge of Fetch API (even less likely in a server-side context, where you have so many other options, and no service worker to deal with)

**Comment by kirill-konshin** [comment]
I have a function that processes the given request -- determines content type (based on headers and contents) and JSON.stringifies or querystring.stringifies (and so on) the request body. The request is provided by consumers of the lib.

**Comment by dbo** [comment]
I'd really would love to see this upstream. We've been using the branch from @kirill-konshin in PR #39 for quite a while without problems and think it's the right way, factoring out Body methods to be spec compliant -- hardly more code, just refactoring. If node-fetch claims to "Stay consistent with window.fetch API." it should fix these issues so users don't run into them, otherwise it's not suitable as a shim for fetch w.r.t. isomorphic code.

Rethink your decision, please.

**Comment by kirill-konshin** [comment]
If package owners decide to say yes, I'll refresh my PR.

**Comment by dbo** [comment]
@bitinn could you please decide on this issue, i.e. either close it or mention how people could help?

**Comment by bitinn** [comment]
@kirill-konshin @dbo back from my vacation, I would say go ahead, refactoring body has been on my TODO list, love to see a new PR tackling it, just `text`, `json` and `buffer` would be enough for now.

but really I don't think we will ever be similar enough to meet the isomorphic requirement.

**Comment by Klortho** [comment]
Hi! Just a data point: I bumped into this problem right away as a naive user, trying to use node-fetch to access a JSON API, because Request.json isn't implemented. It's taken me a long time to find my way to this ticket, and IMO, writing a simple fetch-based JSON scraper should be drop-dead simple. I'd really appreciate it if you could pull one of the solutions to this problem!

**Comment by kirill-konshin** [comment]
Ok, I'll refresh the PR and let you know when I'm done.

**Comment by Klortho** [comment]
Actually, I think I was confused: my problem is that I was calling .json() twice on the same object -- not sure if that's supposed to be permitted or not, but my problem went away when I changed that. Sorry for the noise.

**Comment by kirill-konshin** [comment]
You need to clone your response in order to call `json()` twice. Clone is not yet supported too: see https://github.com/bitinn/node-fetch/issues/34.

**Comment by kirill-konshin** [comment]
I have rebased the PR https://github.com/bitinn/node-fetch/pull/39 on top of most recent changes. But CI fails due to:

```
/home/travis/build/bitinn/node-fetch/node_modules/coveralls/bin/coveralls.js:18
        throw err;
              ^
Bad response: 500 {"message":"Build processing error.","error":true,"url":""}
```

I have nothing to do with it...

**Comment by bitinn** [comment]
this is now in master, I will need to find time to fix coverage and publish a new version.

**Comment by dbo** [comment]
Thanks a lot @bitinn! Excuse my ignorance, what's still missing for a new version? Travis is green.

**Comment by bitinn** [comment]
@dbo there are a few PR I would like to merge and I am short on time at the moment, so no release so far.

**Comment by bitinn** [comment]
released as 1.4.0
